### ISSUE TITLE
Exception in 'threadWait' after a while

### ISSUE THREAD

**Issue by ebergq** [yesodweb/wai#739]
I'm getting `threadWait: invalid argument (Bad file descriptor)` after a while when an endpoint has been triggered.

### Example code
```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}

module Main where

import           Data.Aeson (ToJSON)
import           Data.Text as T
import           GHC.Generics (Generic)
import           Network.Wai.Handler.Warp (run)
import           Servant (Application, Proxy(..), Server, serve)
import           Servant.API ((:>), Get, JSON)

type ModuleAPI = "modules" :> Get '[JSON] [Module]

data Module = Module { moduleName :: T.Text }
  deriving (Eq, Show, Generic)

instance ToJSON Module

mkModule :: T.Text -> Module
mkModule name = Module { moduleName = name }

modules :: [Module]
modules = [ mkModule "module#1" ]

server :: Server ModuleAPI
server = return modules

moduleAPI :: Proxy ModuleAPI
moduleAPI = Proxy

app :: Application
app = serve moduleAPI server

main :: IO ()
main = run 8080 app
```

### Steps to reproduce
1. Compile with profiling options
2. Start the application with `+RTS -xc`
3. Make a request against `http://localhost:8080/modules`
4. Wait for the exception stack trace to be printed to the console

### StackTrace
```
*** Exception (reporting due to +RTS -xc): (base:GHC.Exception.Type.SomeException), stack trace:
  Network.Wai.Handler.Warp.Run.serveConnection.\,
  called from Network.Wai.Handler.Warp.Recv.receiveloop,
  called from Network.Wai.Handler.Warp.Recv.receive.\,
  called from Network.Wai.Handler.Warp.Recv.receive,
  called from Network.Wai.Handler.Warp.Run.socketConnection,
  called from Network.Wai.Handler.Warp.Run.runSettingsSocket.getConn,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnection.getConnMaker,
  called from Network.Wai.Handler.Warp.Run.acceptConnection.acceptNewConnection,
  called from Network.Wai.Handler.Warp.Run.wrappedRecv,
  called from Network.Wai.Handler.Warp.Types.mkSource,
  called from Network.Wai.Handler.Warp.Types.readSource,
  called from Network.Wai.Handler.Warp.Request.headerLines,
  called from Network.Wai.Handler.Warp.Request.recvRequest,
  called from Network.Wai.Handler.Warp.Run.serveConnection.http1,
  called from Network.Wai.Handler.Warp.Run.serveConnection,
  called from Network.Wai.Handler.Warp.Run.fork.serve.\.\,
  called from Network.Wai.Handler.Warp.Run.fork.serve.\,
  called from Network.Wai.Handler.Warp.Run.fork.serve,
  called from Network.Wai.Handler.Warp.Run.fork.\.\,
  called from Network.Wai.Handler.Warp.Run.fork.withClosedRef,
  called from Network.Wai.Handler.Warp.Run.fork.\,
  called from Network.Wai.Handler.Warp.Settings.defaultSettings,
  called from Network.Wai.Handler.Warp.Settings.settingsFork,
  called from Network.Wai.Handler.Warp.Run.fork,
  called from Network.Wai.Handler.Warp.Run.acceptConnection.acceptLoop,
  called from Network.Wai.Handler.Warp.Run.acceptConnection,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withII0.\.\.\.\,
  called from Network.Wai.Handler.Warp.FileInfoCache.withFileInfoCache,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withII0.\.\.\,
  called from Network.Wai.Handler.Warp.FdCache.withFdCache,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withII0.\.\,
  called from Network.Wai.Handler.Warp.Date.withDateCache,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withII0.\,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withTimeoutManager,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure.withII0,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMakerSecure,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnectionMaker,
  called from Network.Wai.Handler.Warp.Run.runSettingsConnection,
  called from Network.Wai.Handler.Warp.Run.runSettingsSocket,
  called from Network.Wai.Handler.Warp.Run.runSettings.\,
  called from Network.Wai.Handler.Warp.Run.runSettings,
  called from Network.Wai.Handler.Warp.Run.run,
  called from Main.main
```

When adding a logger of some kind one can verify that the exception is `threadWait: invalid argument (Bad file descriptor)`. Is there anything I can do about this?

I created this over at servant but I was recommended to open an issue here too. https://github.com/haskell-servant/servant/issues/1165

**Comment by snoyberg** [comment]
Please attempt to reduce this to a repro against Warp itself.

**Comment by t3hmrman** [comment]
I don't have a warp-only repro just yet, but where I think the problem is:

- https://github.com/yesodweb/wai/blob/8773fb48a436fad5f451dbb6a8750fab66d1aeff/warp/Network/Wai/Handler/Warp/Run.hs#L359
- https://github.com/yesodweb/wai/blob/8773fb48a436fad5f451dbb6a8750fab66d1aeff/warp/Network/Wai/Handler/Warp/Run.hs#L553

Just spitballing here but I *think* that the `recv` (aka `connRecv`) is being called on a *closed `Connection` here.

**Comment by kazu-yamamoto** [comment]
I would like to understand the point correctly. Idle connections are cleaned up after 30 seconds or so. What is an expected behavior? Should `ThreadKilled` returned instead of `Bad file descriptor`?

**Comment by t3hmrman** [comment]
Hi @kazu-yamamoto Well I'm not sure what the original poster of this issue wanted for behavior, but personally I'd like this to *not* be an exception if possible, since it's not an "exceptional" case. If a connection is idle and cleaned up, it would be nice to prevent receiving on it without letting the failure happen -- a well placed MVar or IORef seems like it might help here, or even more simply a check to see if the connection is open/closed prior to trying to `connRecv` on it.

Looking at Warp's code the [`Connection` type comments say `connRecv` should return "" for EOF](https://github.com/yesodweb/wai/blob/8773fb48a436fad5f451dbb6a8750fab66d1aeff/warp/Network/Wai/Handler/Warp/Types.hs#L102), maybe that might be a decent behavior though EOF and bad file descriptors are definitely distinct cases.

Maybe warp could look to mirror the code for reading files in `base`? While it throws exceptions too they use this function [`withHandle` and some others](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.IO.Handle.Internals.html#withHandle) to check invariants before use of the handles. While I do not like that `readFile` is written to throw & catch exceptions (I recently started an effort to write a `readFile` that was exception free since I use `+RTS -xc -RTS`in production), I want to note that their exceptions are of a somewhat different kind -- they're using exceptions to exit code blocks (w/ some indication of where they exited and why), more so than because of misuse of some function. Warp's use of exceptions here seems to be less intentional signalling in this manner but more so a use of `recv` (aka `connRecv`) which was unprotected.

Apologies, still have not gotten around to making a warp-only reproduction of this issue.

**Comment by kazu-yamamoto** [comment]
Would you tell me the GHC version and which OS.

**Comment by kazu-yamamoto** [comment]
GHC 8.8.1 on macOS 10.15.2: I can reproduce this with the following code:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import Network.Wai
import Network.HTTP.Types (status200)
import Network.Wai.Handler.Warp

main :: IO ()
main = runSettings settings application
  where
    settings = setOnException (\_ e -> print e) $ setTimeout 2 $ setHost "127.0.0.1" $ setPort 8080 $ defaultSettings
    application _ respond = respond $ responseLBS status200 [("Content-Type", "text/plain")] "Hello World"
```

Client side:

```
% gtelnet 127.0.0.1 8080            
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
Connection closed by foreign host.
```

Server side:

```
% runghc -threaded wai/wai-simple.hs
threadWait: invalid argument (Bad file descriptor)
```

**Comment by kazu-yamamoto** [comment]
I have implemented several toy servers to reproduce this.
While they try to receive data, they are killed by other threads.
In all cases, `ThreadKilled` is caught instead of `Bad file descriptor`.

The warp code tells me that a server thread is killed and then its socket is closed. I don't understand why a server thread can survive and read the socket after its the socket is closed.

@snoyberg Any thoughts?

**Comment by kazu-yamamoto** [comment]
Ah. `closeConn` is also called the timeout manager! That's why.

**Comment by snoyberg** [comment]
Nice, good catch!

**Comment by kazu-yamamoto** [comment]
I'm reading #602 and #605.

**Comment by kazu-yamamoto** [comment]
`setOnException` was introduced in 3.1.0. #605 was merged for 3.2.10.
Isn't `setOnException` good enough for #602? I don't understand why we merged #605 instead of recommending usage of `setOnException`.
